util/functions.js

```diff
@@ -0,0 +1,63 @@
+const flatCache = require('flat-cache');
+const endpointParams = require('./endpointParams.js');
+
+const cache = flatCache.load('rootstock-self-service-support');
+
+function getCacheKey(req) {
+  return (req.baseUrl + req.path).replace(/^\/|\/$/g, '');
```

Comment:
A cleaner way to select the cache key would be to use `req.path` plus each of the keys and their values from `queryStringParams` config for that particular `req.path`.
otherwise a malicious client may perform cache-busting DOS attacks by adding on additional (unused) query params in their requests.

Reply:
I want the endpoint route of type `api/v1/rsk-activity-report/avg-tx-cost` to be a cache key. Cache doesn't know about query string params. It stores only the data returned by the DB. Any additional QS params don't affect cache because only params listed in `enpointParams.js` files are being verified and handled. Other params are just ignored.

Reply:
do you propose some other cache mechanism?

Reply:
so it must store every combination of params and corresponding db responses. Right? For example, chain=mainnet, blocks=100 => cached DB response value for these particular numbers should be stored separately? Now it only stores different values for different chains

Reply:
OK, I resolved this issue. Now the cache keys look as follows: 
```javascript
{
  '/api/v1/rsk-activity-report/avg-tx-cost?chain=rsk_mainnet&blocks=101': {
    time: 2022-12-01T18:56:44.261Z,
    avg_tx_cost_rbtc: 0.000006074137416555041,
    avg_tx_cost_usd: 0.10420055181214426
  },
  '/api/v1/rsk-activity-report/avg-tx-cost?chain=rsk_testnet&blocks=101': {
    time: 2022-12-01T18:57:38.746Z,
    avg_tx_cost_rbtc: 0.000005552545392078367,
    avg_tx_cost_usd: 0.09525275016657204
  },
  '/api/v1/rsk-activity-report/avg-tx-cost?chain=rsk_testnet&blocks=102': {
    time: 2022-12-01T18:59:28.417Z,
    avg_tx_cost_rbtc: 0.000005552545392078367,
    avg_tx_cost_usd: 0.09525275016657204
  },
  '/api/v1/rsk-activity-report/avg-tx-cost?chain=rsk_mainnet&blocks=102': {
    time: 2022-12-01T19:00:40.258Z,
    avg_tx_cost_rbtc: 0.0000060604251924293326,
    avg_tx_cost_usd: 0.1039653214868348
  }
}
```

Reply:
this URL used as a key is not taken from the request query string, but composed from the QS params specified in config file. So it's safe in terms of DOS attacks

Reply:
yep, this is perfect - you definitely need multiple cache entries when they are different query params because the values that are computed are expected to be different.

now there's is a more challenging task: as the number of different types of queries grows, e.g. imagine if someone queries with `blocks=100`,  `blocks=101`, ... all the way to  `blocks=999`;
in this case you will have that many entries in your cache.
a malicious actor, if they figure this out, can DOS this server by causing it to run out of memory using just the allowed params. this is another type of cache busting.

the solution to this is cache invalidation. need to think of custom rules/ logic to decide when to delete a cached value. can use TTL (time to live) strategies, or max count (e.g. if max count is 10, and 11th value is cached, 1st value is invalidated, so the count remains at 10).

a secondary solution to this is a further restriction of allowed param input values. right now, in this API, for numb_blocks, any value within a certain min-max range is allowed. but you could change that by specifying an increment. e.g. instead of 100-1000 so there are 901 possible values. allow only multiples of 100 so that there are only 10 possible values. this way the cache for this particular API will effectively have a max count of 10.